primitives/tracing/src/types.rs

```diff
@@ -0,0 +1,648 @@
+	use std::collections::hash_map::HashMap;
+	use lazy_static::lazy_static;
+
+	// target, name, fieldset, level, kind
+	type MetadataId = (String, String, Vec<String>, WasmLevel, bool);
+
+	lazy_static! {
+		static ref METADATA: Mutex<HashMap<MetadataId, Metadata<'static>>> = Default::default();
+	}
+
+
+	/// Static entry use for wasm-originated metadata.
+	pub struct WasmCallsite;
+	impl callsite::Callsite for WasmCallsite {
+		fn set_interest(&self, _: tracing_core::Interest) { unimplemented!() }
+		fn metadata(&self) -> &Metadata { unimplemented!() }
+	}
+	static CALLSITE: WasmCallsite =  WasmCallsite;
+
+	/// The identifier we are using to inject the wasm events in the generic `tracing` system
+	pub static WASM_TRACE_FALLBACK: &'static str = "wasm_tracing";
+
+	fn mark_static_str<'a>(inp: &'a str) -> &'static str {
+		unsafe {
+			std::mem::transmute(inp)
+		}
+	}
+
+	fn mark_static_slice<'a>(v : &'a[String]) -> &'static [&'static str] {
+		unsafe {
+			std::mem::transmute(v)
+		}
+	}
+
+	fn get_metadata(
+		target: String,
+		name: String,
+		fields: Vec<String>,
+		level: WasmLevel,
+		is_span: bool
+	) -> &'static Metadata<'static> {
+
+		let n = mark_static_str(name.as_str());
+		let t = mark_static_str(target.as_str());
+		let f = mark_static_slice(&fields[..]);
+		let kind = if is_span { Kind::SPAN } else { Kind::EVENT };
+		let lvl = match level {
+			WasmLevel::ERROR => Level::ERROR,
+			WasmLevel::WARN => Level::WARN,
+			WasmLevel::INFO => Level::INFO,
+			WasmLevel::DEBUG => Level::DEBUG,
+			WasmLevel::TRACE => Level::TRACE,
+		};
+
+		let mut l = METADATA.lock();
+		let r = l
+			.entry((target, name, fields, level, is_span))
+			.or_insert_with(|| {
+				Metadata::new(
+					n, t, lvl, None, None, None,
+					FieldSet::new(f, tracing_core::identify_callsite!(&CALLSITE)),
+					kind
+				)
+			});
+
+		let m : &'static Metadata<'static> = unsafe {
+			std::mem::transmute(r)
+		};
+		m
+	}
+
+
+	macro_rules! slice_to_values {
+		($fields:expr, $slice:expr, $len:expr ) => {{
+			fn this_transmute<'a>(xs: &'a [(&'a Field, Option<&'a dyn Value>)]) -> &'a [(&'a Field, Option<&'a dyn Value>); $len] {
+				unsafe {
+					std::mem::transmute(xs.as_ptr())
+				}
+			}
+
+			$fields.value_set(this_transmute($slice))
+		}}
+	}
+
+	fn run_converted<F, T>(attrs: crate::WasmEntryAttributes, f: F) -> T
+	where F: Fn(&'static Metadata<'static>, &ValueSet) -> T
+	{
+		let metadata : &Metadata<'static> = attrs.metadata.into();
+		let fields : HashMap<&str, Field> = metadata.fields()
+			.iter()
+			.map(|f| (f.name(), f))
+			.collect();
+		let mut vl = attrs.fields.0;
+
+		let values = vl.iter_mut().filter_map(|(f, v)| {
+			if let Ok(inner) = std::str::from_utf8(&f.0) {
+				if let Some(field) = fields.get(inner) {
+					return Some((field,
+						match v {
+							Some(d) => d.as_value(),
+							None => None
+						}
+					))
+				}
+			}
+			None
+		}).collect::<Vec<_>>();
+
+		let values = match values.len() {
```

Comment:
This is unfortunately necessary because `tracing_core` uses a private-trait to type-enforce that only `'static arrays` of up to 32 elements can be understood as `ValueSet` (there is no other way to instantiate it) – with a lot of ugly unsafe macro magic :( ...